### ISSUE TITLE
Support beta 3.0

### ISSUE THREAD

**Issue by galvesribeiro** [OrleansContrib/OrleansTestKit#76]
Hey folks!

I'm trying to use TestKit with the new beta 3.0 but it is failing:

`System.TypeLoadException : Method 'GetGrain' in type 'Orleans.TestKit.TestGrainFactory' from assembly 'OrleansTestKit, Version=2.3.6.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.`

Is there any plans to make a least a beta package that support it soon? Otherwise will need to drop its usage.

Thanks.

**Comment by seniorquico** [comment]
Hi, @galvesribeiro. You can pull a _quick_ build against 3.0.0-beta1 from MyGet (see the following page for configuration instructions based on your favorite NuGet client):

https://www.myget.org/feed/seniorquico/package/nuget/OrleansTestKit

I don't have any projects even looking at 3.0 yet. Would you mind testing out this build and following up with a 👍 or 👎? Thanks!

Notes:

- The existing test suite passed without any errors or changes.
- There are new methods in `IGrainFactory`. I did not implement these methods, and I will need a bit more time to dig into the use cases.
- The `Microsoft.Orleans.Core.Legacy` package has been removed. I removed the dependency and legacy logger tests from the test/examples project.

**Comment by galvesribeiro** [comment]
Hello @seniorquico thanks for the quick return.

I can confirm that at least the minimum GetGrain can work for now. I wasn't able to try out other stuff yet as this is the first time I'm using this test kit (nice work btw!).

I'll keep you posted. 

Thanks!

**Comment by galvesribeiro** [comment]
Hello @seniorquico just a heads up.

I've noticed the `Silo.OnDeactivate()` is not doing anything.

If I do something like this:

```csharp
        // OnActivateAsync() is called here, and the state is empty.
         var grain = await this.Silo.CreateGrainAsync<MyGrain>(1);
            var obj = new MyObject{...};

             // The internal non-persistent field should contains 1 item after this call
            await grain.AddObject(obj);
            // Should CleanUp
            await Silo.DeactivateAsync(grain);

           // The internal field still have 1 so if I call AddObject() again I would have 2.
            await grain.OnActivateAsync();
```

The internal field should be cleaned up after I call `DeactivateAsync()` but it looks like it isn't working. Idk if it is something that happened before the beta or something new for you...

**Comment by seniorquico** [comment]
That is currently the expected behavior. See #46 for the discussion. In summary, Orleans does not (or at least did not the last time we discussed this) reuse object instances across different activations. Re-activating an instance sounds like an invalid scenario, and one we should not support. We should probably throw an exception in `OnActivateAsync` to make this clear.

The TestKit really provides a mock _grain activation context_, and it's in this specific context the above limitation makes sense. The name `TestKitSilo` and the pattern of using a field/property named `Silo` in fixtures is an unfortunate relic of when we first started this library internally. The TestKit really has never tried to provide a full, mock silo (e.g. can't have more than one grain, can't have more than one activation of a grain, can't use it with non-grains). For a full, mock silo, I'd recommend looking into `TestCluster` in the `Microsoft.Orleans.TestingHost` package.
